[#44] Implement Blueprint Elements

joshuapease opened the pull request:
### Overview:

- Implements all of the items under "Elements" in [Blueprint](https://staging.blueprint.viget.com/).
- Configures Vite to copy SVGs from the `/src/icons/` directory
- Configures Craft with a filesystem and asset volume (so we can test image related features.
- Installs a standalone "[Viget Parts Kit](https://github.com/vigetlabs/craft-parts-kit)" plugin (not 100% certain where we'll go with this, but I'm enjoying breaking up the base module features.

I've left comments on most of the components

joshuapease committed:
[#44] Adds icon macro

Uses `vite-plugin-static-copy` to copy SVGs from src/icons

Macro inlines SVG files into page.

joshuapease committed:
[#44] Adds a button component

joshuapease commented on .ddev/config.yaml:
```diff
@@ -32,7 +32,7 @@ web_extra_exposed_ports:
   # Run vite in a separate process
   - name: 'vite'
-    command: 'npm install && npm run dev'
+    command: 'npm install && npm run build && npm run dev'
```
Ensures that icons are present on start.

maxfenton replied:
Yeah, this works. I had been trying to work around this by either using `src/public` or the [Vite HMR Public Copy plugin](https://github.com/davidwebca/vite-plugin-hmr-public-copy) — this will work, but I think it won't catch additions or changes you make to that folder while working unless you `ddev restart` (since this is in a daemon and not a terminal tab where `R` would reload Vite) — right?

joshuapease replied:
Ah... you know what. I got thrown off by the name. 

I thought it would only copy the `public` directory. But it seems like you can configure it to copy any directory? 

- https://github.com/davidwebca/vite-plugin-hmr-public-copy?tab=readme-ov-file#copy

I was hesitant to use the `public` directory for icons, because on some of our more complex sites, we occasionally need to reference icons from JS. 

I'll put in a ticket to look into this more deeply. 

![CleanShot 2024-10-03 at 09 44 55@2x](https://github.com/user-attachments/assets/d479e847-bbac-4659-8b51-3f61029da054)

joshuapease replied:
Added

- [#49](https://github.com/vigetlabs/craft-site-starter/issues/49)

joshuapease committed:
[#44] Adds a card component

joshuapease committed:
[#44] Install Alpine JS

joshuapease committed:
[#44] Adds Alert Banner Component

joshuapease committed:
[#44] Adds Accordion Component

joshuapease committed:
[#44] Adds Tag Component

joshuapease committed:
[#44] Adds Dialog Component

joshuapease committed:
[#44] Adds video component

joshuapease committed:
[#44] Install `spacecatninja/imager-x-power-pack`

joshuapease committed:
[#44] Adds image components. Also image card component.

joshuapease committed:
[#44] Install standalone Viget Parts Kit

joshuapease committed:
[#44] Adds local filesystem and asset volume

joshuapease commented on .env.example.dev:
```diff
@@ -30,3 +30,6 @@ CRAFT_SECURITY_KEY=
 CRAFT_DISALLOW_ROBOTS=true
+
+# Asset Settings
+ASSET_FILESYSTEM=local
```
In the future, the env var would let you choose s3 or local depending on your environment or preference.

Future ticket for some default AWS setup
- https://github.com/vigetlabs/craft-site-starter/issues/15

joshuapease commented on config/tailwind/buttons.js:
```diff
@@ -0,0 +1,107 @@
+import plugin from 'tailwindcss/plugin'
```
Exact copy of blueprint (only change was to make an ES Module)

joshuapease commented on config/tailwind/dialog.js:
Exact copy of blueprint (only change was to make an ES Module)

joshuapease commented on tailwind.config.js:
```diff
@@ -1,8 +1,47 @@
+import buttons from './config/tailwind/buttons'
```
Mirroring the TW config of blueprint

joshuapease commented on templates/_components/accordion.twig:
```diff
@@ -0,0 +1,48 @@
+{# @var string|null title #}
```
Corresponding Blueprint Component

https://staging.blueprint.viget.com/elements/accordion/

joshuapease commented on templates/_components/alert-banner.twig:
- https://staging.blueprint.viget.com/elements/alert-banner/

Only change was to use Alpine for dismissible variant.

joshuapease commented on templates/_components/button.twig:
- https://staging.blueprint.viget.com/elements/button/

Key differences:
Twig doesn't really have the concept of slots, so the icon and text are props that can be configured.

joshuapease commented on templates/_components/card-icon.twig:
- https://staging.blueprint.viget.com/elements/card/

Key differences:
- As noted in the button component, Twig doesn't have slots, so it's trickier to make a super flexible component.
- I decided to make separate components for the Image & Icon versions of a card. 
- Blueprint didn't directly have an example of "image" based cards, but I figured it was a common enough use case.

joshuapease commented on templates/_components/card-image.twig:
See comment on `card-icon.twig`

joshuapease commented on templates/_components/dialog.twig:
- https://staging.blueprint.viget.com/elements/dialog/alpine/

Key differences:
- No slots
- Modal is triggered with an event rather than having a button tied to the slot``
- The modal needs a unique ID so we can target it with a CustomEvent 

From experience with our modals on MW, there can be scenarios where the button that triggers a modal can't live close to the modal partial.

Alpine makes this pretty easy

```twig
{# Button can live anywhere on the dom #}
<button x-data @click="$dispatch('$dispatch('dialog:open', { id: 'my-dialog' })')"

{{ include('_components/dialog.twig', {
    content,
    id: 'my-dialog',
    hideClose: true,
}) }}
```

joshuapease commented on templates/_components/image-caption.twig:
- https://staging.blueprint.viget.com/elements/image-caption/

We kinda emulate a slot for this component.

It's a bit clunky, but makes it so we don't have to forward our image component props through the image-caption component.

I could be convinced to do something different:
- Prop drilling isn't the worst
- Add props to our image component that wrap it with the figure/figcaption

joshuapease commented on templates/_components/image.twig:
Curious to hear everyone's thoughts on this.

It's using Imager X's power pack, but wraps that abstraction so if you ever change your image plugin, you don't have to rip it out everywhere.

Main goals for this component.
1. Minimize the amount of transform config you need to spread around your project.
2. Reduce the Imager X specific syntax in your project (would make it easier to switch from Imager X if needed).
3. Use TW screen names in your sizes props.

ppimg() is doing some interesting stuff here.

Here's some [further detail on the fillTransforms setting](https://www.spacecat.ninja/blog/power-pack-and-quick-transform-syntax#:~:text=The%20solution%20is,1800%20pixels%20automatically.).

```js

{% set minWidth = 320 %}
{% set maxWidth = 2200 %}

{{ image ? ppimg(
  image: image,
  transform: [
    minWidth,
    maxWidth, // [1.] Generate transforms between 320 and 2200
    {
      ratio: ratio,
    },
  ],
  params: {
    class: cx(class, 'w-full h-auto'),
    loading,
    fillTransforms: 'auto', // [2.] By using fillTransforms 'auto' we fill in between our min/max widths
    autoFillCount: 5, // [3.] Creates 5 transforms evenly spaced between our min/max widths
    allowUpscale: false,
    sizes: sizes,
  },
) }}
```

joshuapease commented on templates/_components/tag.twig:
- https://staging.blueprint.viget.com/elements/tag/

Identical to Blueprint minus the ability to use slots.

joshuapease commented on templates/_components/video.twig:
- https://staging.blueprint.viget.com/elements/video-embed/

Identical to Blueprint

I think over time our clients would benefit from the following:
- lazy load the iframe
- customizable thumbnail / play button

We also might want to consider using a plugin that parses URLs to get the video ID/Embed URL.

joshuapease commented on templates/_macros/icon.twig:
Simple wrapper around the `svg` function in Craft.

Inlines the SVG markup directly into the page.

Provides some icon sizing logic so you don't have to remember to pass `size-xyz` everywhere. Although there is the option to pass `none` as the size and use your own TW classes in unique scenarios.

joshuapease commented on vite.config.js:
```diff
@@ -27,5 +29,10 @@ export default defineConfig(({ command, mode }) => {
       port: HTTP_PORT,
       origin: env.PRIMARY_SITE_URL + ':' + originPort,
     },
+    plugins: [
+      viteStaticCopy({
+        targets: [{ src: 'src/icons/**/*', dest: './assets/icons' }],
+      }),
+    ],
```
This is how we handle icons on MW and it works pretty well.

Open to other suggestions too. I considered putting them in the `public` directory.  But you [loose the ability to bundle SVG with your JS.](https://vitejs.dev/guide/assets#the-public-directory)

joshuapease committed:
[#44] Tweaks during code review

joshuapease commented on templates/_components/image.twig:
```diff
@@ -0,0 +1,96 @@
+  Keep this in sync with your TW's screens config.
+  https://tailwindcss.com/docs/screens
+#}
+{% set screens = {
+  'sm': '(min-width: 640px)',
+  'md': '(min-width: 768px)',
+  'lg': '(min-width: 1024px)',
+  'xl': '(min-width: 1280px)',
+  '2xl': '(min-width: 1536px)',
+} %}
```
On some sites (like MW), we actually convert our TW config to JSON and then load it via PHP so we can get at these values. 

But.... I'm of the impression lately that it a ton of complexity to save us 5-ish lines of code. 

Do screen sizes really change all that much after a project has been started? This might be an acceptable `DUPE`

ten1seven commented on templates/_components/card-icon.twig:
```diff
@@ -0,0 +1,45 @@
+  {% if title or description %}
+    <div class="flex flex-col gap-8">
+      {% if title %}
+        <h2 class="text-xl font-bold">{{ title }}</h2>
```
This H2 should could probably be a variable. I'm also wondering if H3 is a better default if a typical page structure looks like:

```
H1 Page Title
   |- H2 My Card List
      |- H3 Card Title
      |- H3 Card Title
      |- H3 Card Title

joshuapease replied:
Good call! I'll update in my next PR

ten1seven commented on templates/_components/card-image.twig:
```diff
@@ -0,0 +1,50 @@
+  {% if title or description %}
+    <div class="flex flex-col gap-8">
+      {% if title %}
+        <h2 class="text-xl font-bold">{{ title }}</h2>
```
Same question re: heading.

ten1seven reviewed (approved):
Cool!

joshuapease committed:
[#45] Adds home entry type

joshuapease committed:
[#45] Adds simple Rich Text block

- Adds TW class for default site max width

joshuapease committed:
[#45] Adds Card Grid Block
- Uses both image & icon cards

joshuapease committed:
[#45] Adds text with media block

joshuapease committed:
[#45] Rename primaryLink to href. Adds linkText and newWindow fields.

- Update cards & buttons to use new window.

joshuapease committed:
[#45] Adds Call To Action block

joshuapease committed:
[#45] Adds Page Hero block

joshuapease committed:
[#45] Adds advanced / simple CKEditor configs

- Simple is just Bold, Italic, Underline plus links
- Advanced has more options like headings, styles, lists, etc.

joshuapease committed:
[#45] Actually use _blocks/index.twig

Whoops :)

joshuapease committed:
[n/a] Use real TW class

joshuapease committed:
[#45] Adds rough RTE styling

joshuapease committed:
[#45] Adds a simple .container component for setting a site max width.

joshuapease committed:
[#45] Adds a heading block

joshuapease committed:
[#45] Spacing management on page builder

joshuapease committed:
[#45] Block heading level accessibility

joshuapease committed:
[#45] Add heading to card grid block. Make card heading levels dynamic

joshuapease committed:
[#45] More context on container component

joshuapease committed:
[#45] Misc cleanup

joshuapease committed:
[#51] Nav Project Config

The nav has a max depth of 3 levels. This simplifies our markup since it doesn't have to be recursive

joshuapease committed:
[#51] Basic desktop nav.

The desktop and mobile nav markup will be their own partials.

joshuapease committed:
[#51] Adds skip link

joshuapease committed:
[#51] Adds mobile nav

joshuapease committed:
[#51] Adjust focus strategy for desktop nav.

Instead of using a focus trap, we focus on the first element in the
sub nav, but allow you to tab out of the list. When focus
leaves, we pop the subnav.

joshuapease committed:
[#51] Mobile nav polish

joshuapease committed:
[#51] Footer project config

joshuapease committed:
[#51] Footer markup

joshuapease committed:
[#51] Fix tab control edge cases

joshuapease committed:
[#51] Nav styling changes
